CRM/Core/Form.php

```diff
@@ -2388,66 +2398,115 @@ public static function validateMandatoryFields($fields, $values, &$errors) {
+   * of the variants above would be hit).
    *
    * @return NULL|int
+   *
+   * @throws \CRM_Core_Exception
    */
   protected function setContactID() {
-    $tempID = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
-    if (isset($this->_params) && !empty($this->_params['select_contact_id'])) {
-      $tempID = $this->_params['select_contact_id'];
-    }
-    if (isset($this->_params, $this->_params[0]) && !empty($this->_params[0]['select_contact_id'])) {
-      // event form stores as an indexed array, contribution form not so much...
-      $tempID = $this->_params[0]['select_contact_id'];
-    }
+    $requestedContactID = $this->getRequestedContactID();
 
     // force to ignore the authenticated user
-    if ($tempID === '0' || $tempID === 0) {
+    if ($requestedContactID === 0) {
       // we set the cid on the form so that this will be retained for the Confirm page
       // in the multi-page form & prevent us returning the $userID when this is called
       // from that page
       // we don't really need to set it when $tempID is set because the params have that stored
       $this->set('cid', 0);
-      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => (int) $tempID]);
-      return (int) $tempID;
+      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => $requestedContactID]);
+      return (int) $requestedContactID;
     }
 
-    $userID = CRM_Core_Session::getLoggedInContactID();
-
-    if (!is_null($tempID) && $tempID === $userID) {
-      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => (int) $tempID]);
-      return (int) $userID;
-    }
-
-    //check if this is a checksum authentication
-    $userChecksum = CRM_Utils_Request::retrieve('cs', 'String', $this);
-    if ($userChecksum) {
-      //check for anonymous user.
-      $validUser = CRM_Contact_BAO_Contact_Utils::validChecksum($tempID, $userChecksum);
-      if ($validUser) {
-        CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => (int) $tempID]);
-        CRM_Core_Resources::singleton()->addVars('coreForm', ['checksum' => $userChecksum]);
-        return $tempID;
+    if ($requestedContactID === $this->getAuthenticatedContactID()) {
+      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => $requestedContactID]);
+      // Check if this is a checksum authentication.
+      if ($this->getAuthenticatedCheckSumContactID()) {
+        CRM_Core_Resources::singleton()->addVars('coreForm', ['checksum' => CRM_Utils_Request::retrieve('cs', 'String', $this)]);
       }
+      return $requestedContactID;
     }
+
     // check if user has permission, CRM-12062
-    elseif ($tempID && CRM_Contact_BAO_Contact_Permission::allow($tempID)) {
-      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => (int) $tempID]);
-      return $tempID;
+    if ($requestedContactID && CRM_Contact_BAO_Contact_Permission::allow($requestedContactID)) {
```

Comment:
So this only allows permissioning another user by the logged in user. I wondered about that - but since really only the cid in the url can be check-sum authenticated there hypothetical of a checksum user needed permissioned access to another user does not apply

Reply:
wasn't there always the "contribute as someone else"  on front end forms at some point?

Reply:
@seamuslee001 yep - but you can combine that with logged in user permissions, but not checksum authentication permissions from my read on the before code (presumably cos cid has to be in the url)